### PR TITLE
Extract the WordPress reusable commands to a dedicated package

### PR DESCRIPTION
**Pull request by youknowriad**
Related #48457 

## What?

We want the command menu to be available in multiple WP Admin pages (if not all of WP Admin) but we don't want to writing the same commands over and over again. This PR extracts the commands we currently have in the site editor (creating post and pages and navigation in the site editor) into a dedicated `@wordpress/core-commands` package to be reused in multiple places. That package shouldn't be depending on the "edit-site" package. 

At the moment, I didn't add the command menu UI to any new page, but after this PR lands, it's my next thing to tackle.

**Important notes**

 - Right now I'm using hard-coded urls for "site-editor.php" and "post-new.php". We can't rely on the site editor settings anymore to allow these to be tweaked. In a follow-up we should try to find a way to provide a global config to WP-Admin URLs that is independent of any screen (edit-post or edit-site settings won't work here). Open for ideas.
 - I didn't manage to navigate the site editor without full reload. It's either not possible without access to the "history" object used in the site editor or I'm missing something.

## Testing Instructions

1- Open the site editor and focus the page.
2- Click cmd + k.
3- Use the command menu. (no big changes here)

### CODE REVIEW

File: packages/core-commands/src/site-editor-navigation-commands.js [hunk 1]
```diff
@@ -48,30 +46,42 @@ const getNavigationCommandLoaderPerPostType = ( postType ) =>
...
-							canvas:
-								canvasMode === 'edit' ? canvasMode : undefined,
-						} );
+							...extraArgs,
+						};
+						const targetUrl = addQueryArgs(
+							'site-editor.php',
+							args
+						);
+						document.location = targetUrl;
```
Comment (youknowriad) [inline packages/core-commands/src/site-editor-navigation-commands.js]: Here I've tried using `window.history` but failed to trigger the site editor to reload its content. I'm not sure if there's a way to trigger "client site routing" without having access to the "history" object. Anyone knows?

**Review by mcsf — APPROVED** [review]
It makes sense to extract the commands from `edit-site`.

I don't have a strong opinion here: I understand the motivation for creating a new package, but I wouldn't have opposed keeping everything in `commands` at this early stage, until we have a clearer picture of how to organise everything.

**Comment by youknowriad** [comment]
@jasmussen Can you test this PR and specifically, navigating to pages/templates... and do you think it's acceptable for now to do a full reload?

The only solution I can see to this problem is to actually move the "history/router" to a new dedicated package outside of `@wordpress/edit-site` to be able to use it in the commands across pages (post editor, site editor). It's doable but I feel like it's a bit early for that.

**Comment by jasmussen** [comment]
> Can you test this PR and specifically, navigating to pages/templates... and do you think it's acceptable for now to do a full reload?

Sure, took it for a spin. Here's a GIF of navigating to templates, navigation, etc:

site editor

It isn't entirely clear here what I'm supposed to look for in terms of a full page reload. The closest I could see what when I navigated to Templates > Manage all templates, and then seeing a brief white flash when navigating back. Is that the full reload you're referring to? I'm not seeing it in trunk.

If yes, then I have to admit this makes me a little nervous because it's a white flash, and is exactly what a reacty application like this can avoid. Thinking ahead, one of our goals for 6.3 is to make the editor feel really polished, resilient, and solid. Just one trivial aspect of that is to ensure an initial progressbar loading state for the site editor, and unified loading states across all subpages (the skeleton flashing in Navigation for example doesn't seem to match anything we use anywhere else). I know those aspects are orthogonal to what you're asking. But I bring them up because they are subtle details which embrace the appy nature, whereas a white flash, even if brief, is one of the things we want to avoid by adding such loaders.

**Comment by youknowriad** [comment]
No, I didn't touch navigation from the sidebar at all here, I was referring to the navigation from the command menu (cmd + k)

**Comment by jasmussen** [comment]
Apologies, took it for another spin:

another test

I'm not really experiencing anything here that feels off to me. Anything in particular I should be on the look out for?

Only one issue, which seems unrelated to this PR: if I press ⌘K and press Escape, then I can't press ⌘K again until I click somewhere in the dark material. It seems focus is moved to a place where the shortcut doesn't work, perhaps?

**Comment by youknowriad** [comment]
@jasmussen did you try navigating to say a template "index" from the command menu?

**Comment by jasmussen** [comment]
Oh I see the blink now:

index

As I noted when I thought the blink was about something else, I think this is the type of thing to mainly avoid if at all possible. Even if it's a super subtle literally blink and you'll miss it thing, it contributes a little bit to a feeling of being not as considered as it should be. If it's a temporary thing, it seems fine. But ideally the commandbar can be a little aware that it's already in the main section, when it navigates to a sub-section. 

Or to put it even simpler — we mainly want to avoid those blinks. If we can make those blinks invisible or not happen, that's really the main thing.

**Comment by youknowriad** [comment]
I guess in that case, I see two options:

 1- Extract the "router/history" from the site editor. So this means probably a new package (to be honest I considered extracting this before). It can be all private APIs for now.
 2- Duplicate the commands: Use the ones that navigate using the router in the site editor, and use the commands that do full page refresh in other pages.

What do you think @mcsf ? I was thinking of trying the first option but open to suggestions.

**Comment by mcsf** [comment]
> I was thinking of trying the first option but open to suggestions.

I would say that extracting that logic and having consistent router support across contexts is inevitably what we'll need to do sooner or later (also see Phase 3 and WP-Admin modernisation). That said, it's more than just extracting to a new package, right? If I'm not mistaken, this would mean changing from "one page = one editor containing one router" to "one page = one supervisor + router capable of loading different editors", no?

**Comment by youknowriad** [comment]
> "one page = one editor containing one router" to "one page = one supervisor + router capable of loading different editors", no?

Not really, the idea here is to just to have the router API (consuming it from the context essentially) as an independent package. It will still be only used in the site editor and the command could rely on it to navigate in the site editor and fallback to regular navigation in any other wp-admin page.

**Comment by mcsf** [comment]
Ah right, I was getting ahead of myself, thinking about integrating routing/history across our editors. 😅 

Yes, that sounds good. 👍

**Comment by noisysocks** [comment]
> * In a follow-up we should try to find a way to provide a global config to WP-Admin URLs that is independent of any screen (edit-post or edit-site settings won't work here). Open for ideas.

An equivalent to Core's [`admin_url()`](https://developer.wordpress.org/reference/functions/admin_url/) function would be useful. I think it could be as simple as having `script-loader.php` set `wp.url.adminUrlBasePath = 'https://example.com/wp-admin'` in an inline script and then having a `wp.url.adminUrl()` function. I don't see a problem with repeating `'site-editor.php'`, `'edit.php'`, etc. throughout the codebase. It's what Core does. In practice these URLs never change due to BC.

**Comment by youknowriad** [comment]
@jasmussen I extracted the router package so now I restored the frontend navigation (like it is in trunk)

**Comment by jasmussen** [comment]
Need a test?

**Comment by youknowriad** [comment]
Always :P

**Comment by youknowriad** [comment]
I guess I forgot that I checked the "auto-merge" checkbox :P

**Comment by gziolo** [comment]
> > * In a follow-up we should try to find a way to provide a global config to WP-Admin URLs that is independent of any screen (edit-post or edit-site settings won't work here). Open for ideas.
> 
> An equivalent to Core's [`admin_url()`](https://developer.wordpress.org/reference/functions/admin_url/) function would be useful. I think it could be as simple as having `script-loader.php` set `wp.url.adminUrlBasePath = 'https://example.com/wp-admin'` in an inline script and then having a `wp.url.adminUrl()` function. I don't see a problem with repeating `'site-editor.php'`, `'edit.php'`, etc. throughout the codebase. It's what Core does. In practice these URLs never change due to BC.

That would work. We took a similar approach for sharing URLs for nonce or REST API endpoints:

https://github.com/WordPress/wordpress-develop/blob/991fa6bff23590fe71e9f2807cfcdb218d661c95/src/wp-includes/script-loader.php#L342L368

We also hardcoded some paths in the code a long time ago, examples:

https://github.com/WordPress/gutenberg/blob/488295de1d819e98988f20ba3bb69d2c282afa1d/packages/editor/src/components/post-locked-modal/index.js#L153

https://github.com/WordPress/gutenberg/blob/488295de1d819e98988f20ba3bb69d2c282afa1d/packages/edit-post/src/components/browser-url/index.js#L17

**Pull request merged**